In [1]:
import numpy as np

In [2]:
import random

In [3]:
numbers = [random.randint(0, 100) for i in range(100)]

In [4]:
numbers

[80,
 9,
 24,
 69,
 16,
 27,
 70,
 22,
 16,
 65,
 43,
 84,
 28,
 5,
 37,
 78,
 100,
 48,
 7,
 64,
 1,
 58,
 75,
 51,
 26,
 70,
 87,
 34,
 83,
 77,
 71,
 73,
 35,
 35,
 11,
 86,
 90,
 48,
 66,
 82,
 61,
 71,
 76,
 52,
 14,
 31,
 27,
 100,
 74,
 54,
 18,
 85,
 1,
 90,
 46,
 8,
 21,
 10,
 9,
 10,
 50,
 33,
 95,
 88,
 70,
 34,
 19,
 84,
 72,
 66,
 51,
 42,
 19,
 73,
 20,
 73,
 75,
 11,
 65,
 35,
 84,
 25,
 100,
 43,
 68,
 14,
 95,
 55,
 49,
 9,
 5,
 40,
 72,
 49,
 5,
 80,
 15,
 100,
 76,
 40]

## Mean 平均

In [5]:
np.mean(numbers)

50.08

In [6]:
sorted(numbers)

[1,
 1,
 5,
 5,
 5,
 7,
 8,
 9,
 9,
 9,
 10,
 10,
 11,
 11,
 14,
 14,
 15,
 16,
 16,
 18,
 19,
 19,
 20,
 21,
 22,
 24,
 25,
 26,
 27,
 27,
 28,
 31,
 33,
 34,
 34,
 35,
 35,
 35,
 37,
 40,
 40,
 42,
 43,
 43,
 46,
 48,
 48,
 49,
 49,
 50,
 51,
 51,
 52,
 54,
 55,
 58,
 61,
 64,
 65,
 65,
 66,
 66,
 68,
 69,
 70,
 70,
 70,
 71,
 71,
 72,
 72,
 73,
 73,
 73,
 74,
 75,
 75,
 76,
 76,
 77,
 78,
 80,
 80,
 82,
 83,
 84,
 84,
 84,
 85,
 86,
 87,
 88,
 90,
 90,
 95,
 95,
 100,
 100,
 100,
 100]

In [7]:
np.percentile(numbers, 50)

50.5

In [15]:
from scipy import stats

In [16]:
stats.mode(numbers)

ModeResult(mode=array([100]), count=array([4]))

In [13]:
from collections import Counter

In [14]:
Counter(numbers).most_common(1)

[(100, 4)]

+ mean, 平均值 是 50.08
+ median, 中位数 是 50.5
+ mode, 众数 是 100(4)

+ mean, 平均值 是 100
+ median, 中位数 是 20.5
+ mode, 众数 是 30(5)

## 标准差、方差

In [23]:
numbers = [random.randint(0, 1000) for i in range(100)]

In [27]:
std = np.std(numbers)

In [26]:
mean = np.mean(numbers)

## 数据的 Normalization

$$ x_{normalized} = \frac{(x - \bar{X})}{std}  $$

In [36]:
np.std([(n - mean) / std for n in numbers])

1.0

In [37]:
np.mean([(n - mean) / std for n in numbers])

1.4432899320127036e-17

## Outliner

In [38]:
numbers

[640,
 970,
 769,
 991,
 186,
 81,
 886,
 52,
 185,
 166,
 858,
 381,
 581,
 121,
 533,
 626,
 191,
 153,
 879,
 57,
 26,
 353,
 975,
 545,
 2,
 563,
 354,
 479,
 983,
 442,
 249,
 869,
 827,
 90,
 893,
 968,
 19,
 4,
 275,
 968,
 947,
 41,
 338,
 106,
 103,
 500,
 774,
 428,
 824,
 132,
 9,
 392,
 4,
 271,
 820,
 634,
 768,
 822,
 535,
 345,
 113,
 926,
 690,
 229,
 529,
 431,
 153,
 465,
 216,
 919,
 89,
 147,
 887,
 249,
 4,
 569,
 136,
 7,
 509,
 316,
 999,
 244,
 172,
 527,
 485,
 815,
 854,
 7,
 733,
 678,
 369,
 880,
 497,
 155,
 357,
 368,
 606,
 423,
 538,
 608]

In [41]:
np.percentile(numbers, 25) / 1.5

108.83333333333333

In [43]:
[n for n in numbers if n < np.percentile(numbers, 25) / 1.5]

[81, 52, 57, 26, 2, 90, 19, 4, 41, 106, 103, 9, 4, 89, 4, 7, 7]

In [42]:
np.percentile(numbers, 75) * 1.5

1152.375

## Assigments 新华社文章被抄袭判断

In [44]:
filepath = '/Users/mqgao/Workspace/Lecture/datasource/sqlResult_1558435.csv'

In [45]:
import pandas as pd

In [46]:
news = pd.read_csv(filepath, encoding='gb18030')

In [47]:
news.head()

id      author                  source  \
0  89617         NaN  快科技@http://www.kkj.cn/   
1  89616         NaN  快科技@http://www.kkj.cn/   
2  89615         NaN  快科技@http://www.kkj.cn/   
3  89614         NaN                     新华社   
4  89613  胡淑丽_MN7479                   深圳大件事   

                                             content  \
0  此外，自本周（6月12日）起，除小米手机6等15款机型外，其余机型已暂停更新发布（含开发版/...   
1  骁龙835作为唯一通过Windows 10桌面平台认证的ARM处理器，高通强调，不会因为只考...   
2  此前的一加3T搭载的是3400mAh电池，DashCharge快充规格为5V/4A。\r\n...   
3    这是6月18日在葡萄牙中部大佩德罗冈地区拍摄的被森林大火烧毁的汽车。新华社记者张立云摄\r\n   
4  （原标题：44岁女子跑深圳约会网友被拒，暴雨中裸身奔走……）\r\n@深圳交警微博称：昨日清...   

                                             feature  \
0  {"type":"科技","site":"cnbeta","commentNum":"37"...   
1  {"type":"科技","site":"cnbeta","commentNum":"15"...   
2  {"type":"科技","site":"cnbeta","commentNum":"18"...   
3  {"type":"国际新闻","site":"环球","commentNum":"0","j...   
4  {"type":"新闻","site":"网易热门","commentNum":"978",...   

                           title  \
0           小米MIUI 9首批机型曝光：共计15款   
1     骁龙835在Windows 10上的性能表现有望改善   
2      一加手机5细节曝光：3300mAh、充半小时用1天   
3  葡森林火灾造成至少62人死亡 政府宣布进入紧急状态（组图）   
4       44岁女子约网友被拒暴雨中裸奔 交警为其披衣相随   

                                                 url  
0     http://www.cnbeta.com/articles/tech/623597.htm  
1     http://www.cnbeta.com/articles/tech/623599.htm  
2     http://www.cnbeta.com/articles/tech/623601.htm  
3  http://world.huanqiu.com/hot/2017-06/10866126....  
4  http://news.163.com/17/0618/00/CN617P3Q0001875...

In [50]:
from_xinhua_net = news[news['source'] == '新华社']

In [55]:
from_xinhua_net.tail()

id  author source                                            content  \
89606   5      邹峥    新华社  新华社照片，多伦多，2017年6月7日\n（体育）（2）冰球——国家女子冰球队海外选秀在多伦...   
89607   4      王衡    新华社  新华社兰州6月3日电（王衡、徐丹）记者从甘肃省交通运输厅获悉，甘肃近日集中开建高速公路、普通...   
89608   3      张旌    新华社  \n\n2017年5月29日，在法国巴黎郊外的凡尔赛宫，法国总统马克龙出席新闻发布会。（新华...   
89609   2     夏文辉    新华社  \n\n2017年5月25日，在美国马萨诸塞州剑桥市，哈佛大学毕业生在毕业典礼上欢呼。（新华...   
89610   1  王子江 张寒    新华社  　　新华社德国杜塞尔多夫６月６日电题：乒乓女球迷　\n　　新华社记者王子江、张寒\n　　熊老...   

                                                 feature  \
89606  {"type":"冰球","site":"新华社","url":"http://home.x...   
89607  {"type":"宏观经济","site":"新华社","url":"http://home...   
89608  {"type":"其它","site":"新华社","url":"http://home.x...   
89609  {"type":"其它","site":"新华社","url":"http://home.x...   
89610  {"type":"体育","site":"新华社","url":"http://home.x...   

                              title  \
89606  （体育）（2）冰球——国家女子冰球队海外选秀在多伦多展开   
89607         （经济）甘肃集中开工35个重点交通建设项目   
89608          法国议会选举　马克龙有望获“压倒性胜利”   
89609             哈佛大学为何取消这些新生入选资格？   
89610                  （体育）题：乒乓女球迷　   

                                                     url  
89606  http://home.xinhua-news.com/gdsdetailxhsnew/22...  
89607  http://home.xinhua-news.com/gdsdetailxhsnew/22...  
89608  http://home.xinhua-news.com/gdsdetailxhsnew/22...  
89609  http://home.xinhua-news.com/gdsdetailxhsnew/22...  
89610  http://home.xinhua-news.com/gdsdetailxhsnew/22...

In [54]:
len(news)

89611

In [53]:
len(from_xinhua_net) / len(news)

0.8778051801676133

We need build a model which could give the probability of an article is from Xinhua net.

## Notice

1. This must be a machine learning problem. 
2. How to represent the Words Content? 
    + 1. TFIDF: Scikit learning
    + 2. WordEmedding -> Sentence Embedding: Gensim
3. Propercessing, scaling, normalization
4. 
<html>
    <p style='color:red'> Baseline </p>
    <p style='color:red'> Evaluation </p>
 </html>


 5. Try different models, monitor the performance. 


6. Based on the preformance, tuning the parameters. 

Notice : 
    Unbalanced: 1. use F1, F2, precision, recall; 2. if your dataset is enough, we could build a balance sub dataset. 
    